### PR TITLE
[InstCombine] Fold "extract (select (cond, insert(agg, elem), FV))"

### PR DESCRIPTION
**Pull request by weiguozhi**
Do the transform
```
     extract (select (cond, insert(agg, elem), FV))
 ->  select (cond, elem, extract(FV))
```

### CODE REVIEW

**Review by dtcxzyw — APPROVED** [review]
LG. Please wait for additional approval from other reviewers.

We may generalize to use simplifyInstruction for other instructions in the future.

**Comment by nikic** [comment]
> We may generalize to use simplifyInstruction for other instructions in the future.

I'd rather directly do that than add a special case for extractelement. It's easier to generalize this than not. I did a quick test at https://github.com/llvm/llvm-project/pull/116073 and it doesn't seem to affect compile-time, which would have been my main concern.

File: llvm/test/Transforms/InstCombine/extract-select-agg.ll [hunk 1]
```diff
@@ -81,3 +76,25 @@ entry:
...
+; CHECK-NEXT:    [[TMP1:%.*]] = extractvalue { ptr, i64 } [[CALL]], 1
+; CHECK-NEXT:    [[RES:%.*]] = select i1 [[CMP]], i64 4294967294, i64 [[TMP1]]
+; CHECK-NEXT:    ret i64 [[RES]]
+;
+  %call = call { ptr, i64 } @foo()
+  %elm1 = extractvalue { ptr, i64 } %call, 1
+  %cmp = icmp eq i64 %elm1, %v
+  %fca0 = insertvalue { ptr, i64 } poison, ptr %p1, 0
+  %fca1 = insertvalue { ptr, i64 } %fca0, i64 4294967294, 1
+  %select = select i1 %cmp, { ptr, i64 } %fca1, { ptr, i64 } %call
```
[the remark this answers is not in the document]
Comment (weiguozhi) [reply llvm/test/Transforms/InstCombine/extract-select-agg.ll:95]: done.

File: llvm/test/Transforms/InstCombine/extract-select-agg.ll [hunk 2]
```diff
@@ -81,3 +76,25 @@ entry:
...
+  %res = extractvalue { ptr, i64 } %select, 1
+  ret i64 %res
+}
+
```
[the remark this answers is not in the document]
Comment (weiguozhi) [reply llvm/test/Transforms/InstCombine/extract-select-agg.ll:99]: done.

**Comment by dtcxzyw** [comment]
> > We may generalize to use simplifyInstruction for other instructions in the future.
> 
> I'd rather directly do that than add a special case for extractelement. It's easier to generalize this than not. I did a quick test at #116073 and it doesn't seem to affect compile-time, which would have been my main concern.

@nikic Is it okay to just keep these negative tests and make this patch NFC? Positive tests have been added in https://github.com/llvm/llvm-project/commit/2ca25ab11d01ceacf359643b09aed7d53d0ff8dc.

**Comment by nikic** [comment]
> > > We may generalize to use simplifyInstruction for other instructions in the future.
> > 
> > 
> > I'd rather directly do that than add a special case for extractelement. It's easier to generalize this than not. I did a quick test at #116073 and it doesn't seem to affect compile-time, which would have been my main concern.
> 
> @nikic Is it okay to just keep these negative tests and make this patch NFC? Positive tests have been added in [2ca25ab](https://github.com/llvm/llvm-project/commit/2ca25ab11d01ceacf359643b09aed7d53d0ff8dc).

Sure

**Comment by weiguozhi** [comment]
Deprecate this pr because more general version has been committed in https://github.com/llvm/llvm-project/commit/9a844a36eb9a21de27882b6193a82fda49986347.

**Pull request closed without merging**